New spack package: DisCoTec

freifrauvonbleifrei opened the pull request:
This PR introduces a spack package for DisCoTec, an MPI-based framework to run the distributed sparse grid combination technique with various solvers.

freifrauvonbleifrei committed:
new package for DisCoTec

vancraar committed:
change discotec from using scons to cmake

vancraar committed:
change discotec from using scons to cmake

change discotec from using scons to cmake

vancraar committed:
Add build type variant

vancraar committed:
add missing description=

vancraar committed:
Update package.py

vancraar committed:
Update package.py

vancraar committed:
Update package.py

freifrauvonbleifrei committed:
Merge branch 'spack:develop' into develop

freifrauvonbleifrei committed:
DisCoTec package: update flags/versions/defaults

freifrauvonbleifrei committed:
discotec package: format

freifrauvonbleifrei committed:
discotec package: correct cmake flag names

haampie commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,60 @@
+
+    version("main", branch="main")
+
+    variant("enableft", default=False, description="DisCoTec with algorithm-based fault tolerance")
```
`enableft` -> `ft`?

haampie commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,60 @@
+        description="Build type for DisCoTec",
+    )
+
+    depends_on("cmake@3.24.2:", type=("build"))
```
```suggestion
    depends_on("cmake@3.24.2:", type="build")
```

haampie commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,60 @@
+    variant("vtk", default=False, description="Build with VTK support")
+    variant("lto", default=True, description="Build with link-time optimization")
+    variant(
+        "build_type",
```
this is defined by default, does the package not support all default options?

freifrauvonbleifrei replied:
yes, actually it does!

haampie commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,60 @@
+    depends_on("highfive+mpi+boost+ipo", when="+hdf5")
+    depends_on("vtk", when="+vtk")
+
+    def cmake_args(self):
```
Use `self.define_from_variant(<option>, <variant>)` below, it's easier.

freifrauvonbleifrei replied:
this is a great suggestion, thx!

freifrauvonbleifrei committed:
discotec package: add selalib flag

freifrauvonbleifrei committed:
discotec package: use some review suggestions

freifrauvonbleifrei committed:
discotec package: update string delimiters
(to make style checker happy, although the docs list examples with single quotes -- https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html )

freifrauvonbleifrei committed:
discotec package: remove unused cmake flag

freifrauvonbleifrei commented:
is there something I can do to help this PR move forward? :)

freifrauvonbleifrei commented:
@haampie @alalazo ?

freifrauvonbleifrei committed:
discotec package.py : default-enable OpenMP

alalazo commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,55 @@
+    variant("lto", default=True, description="Build with link-time optimization")
+    variant("openmp", default=True, description="Parallelize with OpenMP")
+    variant("timing", default=True, description="With high-res timers")
+    variant("test", default=True, description="Build Boost tests")
```
What is the purpose of this variant? Can we get rid of it and use `self.run_tests` below?

freifrauvonbleifrei replied:
it switches whether the tests are built or not -- do you think this would better be part of run_tests?

alalazo replied:
Yes, I think it would be better using `self.run_tests` if it is possible to do so.

alalazo commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,55 @@
+    depends_on("cmake@3.24.2:", type="build")
+    depends_on("glpk")
+    depends_on("highfive+mpi+boost+ipo", when="+hdf5")
+    depends_on("mpi", type=("build", "run"))
```
Why `mpi` is of type `build,run`? Don't you `link` to it?

freifrauvonbleifrei replied:
yes, but the mpi compilers need to be used for building, and `mpirun` (or similar) needs to be used for executing (and the same mpi versions at that). is `link` something I should add to the list?

alalazo replied:
I would use the default and remove `type=` alltogether. That's what most `mpi` software do.

alalazo commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,55 @@
+    variant("vtk", default=False, description="Build with VTK support")
+
+    depends_on(
+        "boost +test +serialization +filesystem +system +program_options", type=("build", "run")
```
Same question here for the `build,run` type.

freifrauvonbleifrei replied:
hm, I assume the flags could be removed. let me try ;)

alalazo reviewed (changes requested):
A few questions, but generally LGTM.

freifrauvonbleifrei committed:
DisCoTec package.py: update dependencies

freifrauvonbleifrei committed:
DisCoTec package.py: reformat boost dependency line

freifrauvonbleifrei committed:
Merge branch 'develop' into develop

freifrauvonbleifrei committed:
Update discotec: package.py

according to PR feedback

freifrauvonbleifrei committed:
Merge branch 'develop' into develop

freifrauvonbleifrei committed:
Merge branch 'develop' into develop

alalazo commented:
@haampie Are you fine with this one?

freifrauvonbleifrei committed:
Merge branch 'develop' into develop

alalazo commented on var/spack/repos/builtin/packages/discotec/package.py:
```diff
@@ -0,0 +1,51 @@
+    url = "https://github.com/SGpp/DisCoTec"
+    git = "https://github.com/SGpp/DisCoTec"
+
+    # notify when the package is updated.
+    maintainers = ["freifrauvonbleifrei", "pfluegdk"]
```
```suggestion
    maintainers("freifrauvonbleifrei", "pfluegdk")
```

alalazo committed:
Use "maintainers" directive
